jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java

```diff
@@ -104,15 +128,285 @@ public Object getAttribute(String name)
         {
             _attributes.clearAttributes();
         }
+
+        @Override
+        public int hashCode()
+        {
+            return _attributes.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return _attributes.equals(obj);
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by a {@link ConcurrentHashMap}.
+     */
+    class Mapped implements Attributes
+    {
+        private final java.util.concurrent.ConcurrentMap<String, Object> _map = new ConcurrentHashMap<>();
+        private final Set<String> _names = Collections.unmodifiableSet(_map.keySet());
+
+        public Mapped()
+        {
+        }
+
+        public Mapped(Mapped attributes)
+        {
+            _map.putAll(attributes._map);
+        }
+
+        @Override
+        public Object removeAttribute(String name)
+        {
+            return _map.remove(name);
+        }
+
+        @Override
+        public Object setAttribute(String name, Object attribute)
+        {
+            if (attribute == null)
+                return _map.remove(name);
+
+            return _map.put(name, attribute);
+        }
+
+        @Override
+        public Object getAttribute(String name)
+        {
+            return _map.get(name);
+        }
+
+        @Override
+        public Set<String> getAttributeNames()
+        {
+            return _names;
+        }
+
+        @Override
+        public void clearAttributes()
+        {
+            _map.clear();
+        }
+
+        public int size()
+        {
+            return _map.size();
+        }
+
+        public Set<Map.Entry<String, Object>> getAttributes()
+        {
+            return _map.entrySet();
+        }
+
+        @Override
+        public String toString()
+        {
+            return _map.toString();
+        }
+
+        public void addAll(Attributes attributes)
+        {
+            for (String name : attributes.getAttributeNames())
+                setAttribute(name, attributes.getAttribute(name));
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return _map.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (o instanceof Attributes)
+            {
+                Attributes a = (Attributes)o;
+                for (Map.Entry<String, Object> e : _map.entrySet())
+                {
+                    if (!Objects.equals(e.getValue(), a.getAttribute(e.getKey())))
+                        return false;
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by another Attributes instance, which is treated as immutable, but with a
+     * ConcurrentHashMap used as a mutable layer over it.
+     */
+    class Layer implements Attributes
+    {
+        private static final Object REMOVED = "REMOVED";
```

Comment:
```java
private static final Object REMOVED = new Object();
```

Reply:
The benefit of using a String is that it dumps well.  I could use an Object, but then would need to override the toString so it dumps/debugs well.

Reply:
The drawback is that using this special string as a value is exploitable, either or purpose or by accident.